### PR TITLE
luanaruiz9 test rel attn copy decoder

### PR DESCRIPTION
**Pull request by luanaruiz9**
Santi, Josh,

Can you please have a look at my implementation of relative attention and of the copy decoder? The main changes are in models.py. The changes in train.py and in the configs file refer to new config variables/flags and in input_pipeline.py I disabled dataset packing in the training set, as I believe that would interfere with the current implementation of relative attention.

Presently, I am not able to properly learn models with relative attention. I believe that there is something wrong with this implementation as training stalls at around 30% token-level accuracy and 0% sentence-level accuracy. 

Here is an example of error (with relative attention only, relative radius 16) that I thought was curious as most of the tokens are correct but the order is wrong.

remove_first echo G16 L3 E5 W8 , T3 A8 E3 (input)

remove_first G16 L3 E5 W8 W8 , T3 A8 E3 (target)

remove_first A8 T3 W8 , E5 G16 E3 L3 (prediction)

As for the copy decoder, the biggest difference with the keras implementation is that the softmax of the logits is not calculated in the transformer (models.py), but in train.py, so it is the output of the decoder and of the copy decoder that are being mixed.

Thanks!

### CODE REVIEW

File: model/models.py [hunk 1]
```diff
@@ -89,6 +106,38 @@ def init(key, shape, dtype=np.float32):
     return jnp.array(pe)
 
   return init
+  
+
+def create_relative_ids(in_length, relative_radius, tar_length=None, de2enc_ids=None):
```
Comment (santiaontanon-google) [inline model/models.py:111]: typo: "de2enc_ids" (this is my original typo, hehehe)

File: model/models.py [hunk 2]
```diff
@@ -89,6 +106,38 @@ def init(key, shape, dtype=np.float32):
...
+      
+  if tar_length is not None:
+      dec_relative_ids = jnp.zeros([tar_length, tar_length], dtype=int)
+      for i in range(tar_length):
+        for j in range(tar_length):
+          diff = i - j
+          diff = relative_radius + min(max(diff, -relative_radius), relative_radius)
+          dec_relative_ids.at[i,j].set(diff)
+    
+      dec2end_relative_ids = jnp.zeros([tar_length, in_length], dtype=int)
```
Comment (santiaontanon-google) [inline model/models.py:128]: nit: "dec2end" -> "dec2enc" (also my original typo, as it should mean "DECoder to ENCoder" :) )

File: model/models.py (hunk 1 above)
Comment (santiaontanon-google) [inline model/models.py:111]: I wonder if this is the function that is to blame for the slow downs you mentioned. JAX is functional, which means that each time we assign a value to an array, the array is cloned, and a new one with the updated value is created. This code was written for numpy/Python. So perhaps you should use "np" instead of "jnp", and only convert to "jnp" in the return statement?

File: model/models.py [hunk 3]
```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
...
+  Args:
+    query: queries for calculating attention with shape of
+      `[batch..., q_length, num_heads, qk_depth_per_head]`.
+    key: keys for calculating attention with shape of
+      `[batch..., kv_length, num_heads, qk_depth_per_head]`.
+    bias: bias for the attention weights. This should be broadcastable to the
+      shape `[batch..., num_heads, q_length, kv_length]`.
+      This can be used for incorporating causal masks, padding masks,
+      proximity bias, etc.
+    broadcast_dropout: bool: use a broadcasted dropout along batch dims.
```
Comment (santiaontanon-google) [inline model/models.py:265]: relative_ids, relative_embeddings, relative_biases are missing form the docstring.

File: model/models.py [hunk 4]
```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
...
+  assert query.ndim == key.ndim, 'q, k must have same rank.'
+  assert query.shape[:-3] == key.shape[:-3], (
+      'q, k batch dims must match.')
+  assert query.shape[-2] == key.shape[-2], (
+      'q, k num_heads must match.')
+  assert query.shape[-1] == key.shape[-1], 'q, k depths must match.'
+
+  # calculate attention matrix
+  depth = query.shape[-1]
+  query = query / jnp.sqrt(depth).astype(dtype)
```
Comment (santiaontanon-google) [inline model/models.py:285]: I don't think this will make much of a difference, but if added here, the "relative_biases" will not be affected by this division. So, maybe you want to move this to later, and directly divide the attn_weights, like "attn_weights = attn_weights / jnp.sqrt(depth).astype(dtype)" after the relative_biases have been added.

File: model/models.py [hunk 5]
```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
...
+      `[batch..., q_length, num_heads, qk_depth_per_head]`.
+    key: keys for calculating attention with shape of
+      `[batch..., kv_length, num_heads, qk_depth_per_head]`.
+    value: values to be used in attention with shape of
+      `[batch..., kv_length, num_heads, v_depth_per_head]`.
+    bias: bias for the attention weights. This should be broadcastable to the
+      shape `[batch..., num_heads, q_length, kv_length]`.
+      This can be used for incorporating causal masks, padding masks,
+      proximity bias, etc.
+    broadcast_dropout: bool: use a broadcasted dropout along batch dims.
```
Comment (santiaontanon-google) [inline model/models.py:354]: nit: also missing arguments in the docstring

File: model/models.py [hunk 6]
```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
...
+        for details.
+      kernel_init: initializer for the kernel of the Dense layers.
+      bias_init: initializer for the bias of the Dense layers.
+      use_bias: bool: whether pointwise QKVO dense transforms use bias.
+      attention_fn: dot_product_attention or compatible function. Accepts
+        query, key, value, and returns output of shape
+        `[bs, dim1, dim2, ..., dimN,, num_heads, value_channels]``
+      decode: whether to prepare and use an autoregressive cache.
+  """
+  num_heads: int
```
Comment (santiaontanon-google) [inline model/models.py:407]: Maybe it's cleaner to have a "config: TransformerConfig" here instead of all the individual variables?

File: model/models.py [hunk 7]
```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
...
+  precision: Any = None
+  kernel_init: Callable[[PRNGKey, Shape, Dtype], Array] = default_kernel_init
+  bias_init: Callable[[PRNGKey, Shape, Dtype], Array] = zeros
+  use_bias: bool = True
+  attention_fn: Callable[[Array, Array, Array], Array] = dot_product_relative_attention
+  decode: bool = False
+  relative_radius: Optional[int] = None
+  relative_bias: bool = False
+  
+  def setup(self):
```
Comment (santiaontanon-google) [inline model/models.py:423]: Hmm, I don't know if it makes a difference. But when I implemented relative attention, I got JAX errors if I initialized the "nn.Embed" layers in the "setup" method. I just create them in the __call__ method, like the other layers.

**Review by santiaontanon-google — COMMENTED** [review]
Hi Luana, I only looked at the relative attention code (didn't look at the copy decoder yet). I only found a couple of nits here and there, that I don't think would make much of a difference, perhaps the one thing is the place where you initialize the nn.Embed layers, that's worth a try! Other than that, form the results you report, it seems like position embeddings are not added, so, is there any easy way in which you can actually verify the relative attention code is being called and the relative_ids make it all the way to the attention code? (just wondering if they are None-ed, and some point due to some typo or something.

Another general comment is that instead of duplicating the attention functions, it might have been easier to leave the original functions/classes, but just have the optional relative parameters, which if None are ignored, and no relative atention occurs. But let's get it working first, and we can worry about code cleanness later :)

File: model/models.py [hunk 8]
```diff
@@ -20,14 +20,26 @@
 # pytype: disable=wrong-keyword-args
 # pytype: disable=attribute-error
 
-from typing import Callable, Any, Optional
+from functools import partial
+from typing import (Any, Callable, Tuple, Optional)
```
Comment (jtainslie) [inline model/models.py:24]: Nit: parentheses can be removed.
```suggestion
from typing import Any, Callable, Tuple, Optional
```

File: model/models.py [hunk 9]
```diff
@@ -20,14 +20,26 @@
...
 from flax import struct
 from jax import lax
+from jax import random
 import jax.numpy as jnp
 import numpy as np
 
+from flax.linen.linear import default_kernel_init
+from flax.linen.linear import DenseGeneral
+from flax.linen.module import Module, compact, merge_param
+from flax.linen.initializers import zeros
```
Comment (jtainslie) [inline model/models.py:37]: Nit: imports should be for modules/packages, not classes/functions.
https://google.github.io/styleguide/pyguide.html#22-imports

File: model/models.py [hunk 10]
```diff
@@ -50,6 +62,11 @@ class TransformerConfig:
   kernel_init: Callable = nn.initializers.xavier_uniform()
   bias_init: Callable = nn.initializers.normal(stddev=1e-6)
   posemb_init: Optional[Callable] = None
+  sinusoidal: bool = False
+  relative_radius: Optional[int] = None
+  relative_bias: Optional[bool] = False
+  enc2dec: Optional[bool] = False
+  copy_decoder: Optional[bool] = None
```
Comment (jtainslie) [inline model/models.py:69]: Nit: Just checking if `None` has a different role than `False` in these.  If not, it'd be simpler to make them `bool` instead of `Optional[bool]`.

File: model/models.py (hunk 1 above)
Comment (jtainslie) [inline model/models.py:111]: Nit: This function could probably benefit from a docstring and a unit test.

File: model/models.py [hunk 11]
```diff
@@ -89,6 +106,38 @@ def init(key, shape, dtype=np.float32):
...
+  
+
+def create_relative_ids(in_length, relative_radius, tar_length=None, de2enc_ids=None):
+    
+  enc_relative_ids = jnp.zeros([in_length, in_length], dtype=int)
+  for i in range(in_length):
+    for j in range(in_length):
+      diff = i - j
+      diff = relative_radius + min(max(diff, -relative_radius), relative_radius)
+      enc_relative_ids.at[i,j].set(diff)
```
Comment (jtainslie) [inline model/models.py:118]: Nit: instead of setting each value individually, you could "vectorize" this operation by broadcasting the subtraction of a column and row vector.  (Same with the cases below.)  This could speed things up in JAX too regarding Santi's comment above.

File: model/models.py [hunk 12]
```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
...
+  def setup(self):
+    self.relative_embeddings = None
+    self.relative_biases = None
+    if self.relative_radius is not None:
+      relative_vocab_size = 2 * self.relative_radius + 1
+      head_dim = self.qkv_features // self.num_heads
+      self.relative_embeddings = nn.Embed(relative_vocab_size, head_dim, 
+                                          embedding_init=nn.initializers.normal(stddev=1.0))
+      if self.relative_bias:
+        self.relative_biases = nn.Embed(relative_vocab_size, 1,
```
Comment (jtainslie) [inline model/models.py:432]: Nit: In the general case, there can be a different bias per head, which could potentially add some expressiveness at very little cost.

File: model/models.py [hunk 13]
```diff
@@ -578,10 +1010,28 @@ def __call__(self,
...
+      copy_output_weight = nn.sigmoid(self.final_layer_copy_weight(dec_output))
+      copy_output = dot_product_relative_attention(
+          copy_output_query, # (batch_size, tar_seq_len, d_model)
+          encoded,         # (batch_size, inp_seq_len, d_model)
+          jax.nn.one_hot(inputs, cfg.output_vocab_size)) # (batch_size, inp_seq_len, vocab_size)
+      if cfg.qkv_dim < cfg.output_vocab_size:
+          copy_output = self.final_layer_copy2b(self.final_layer_copy2a(copy_output))
+      else:
+          copy_output = self.final_layer_copy2(copy_output)
+      final_output = (1-copy_output_weight) * final_output + copy_output_weight * copy_output
```
Comment (jtainslie) [inline model/models.py:1035]: Formatting nit:
```suggestion
      final_output = (1 - copy_output_weight) * final_output + copy_output_weight * copy_output
```

File: model/models.py (hunk 3 above)
Comment (jtainslie) [reply model/models.py:265]: It's helpful to make sure the docstrings for any tensors (e.g. `relative_ids`) include their shapes too!

File: model/models.py [hunk 14]
```diff
@@ -186,17 +235,351 @@ def __call__(self, inputs):
...
+  depth = query.shape[-1]
+  query = query / jnp.sqrt(depth).astype(dtype)
+  # attn weight shape is (batch..., num_heads, q_length, kv_length)
+  attn_weights = jnp.einsum('...qhd,...khd->...hqk', query, key,
+                            precision=precision)
+
+  if relative_ids is not None:
+    if relative_embeddings is not None:
+        r = relative_embeddings(relative_ids)
+        matmul_qrel = jnp.einsum('...qhd,...qkd->hqk', query, r,
```
Comment (jtainslie) [inline model/models.py:293]: Looks like the `hqk` output of the einsum formula is dropping the `...` batched dimension, which is probably not what we want right?

**Review by jtainslie — APPROVED** [review]
Lgtm apart from some nits and a potential adjustment to an einsum formula.
